Fixing policy delete root

aanm opened the pull request:
Fixes #410 #411 
The commits have their own descriptions. I can't split the commits into multiple PRs since they are slightly tight together.

I ran the test from #436  (start 10 containers, wait them to be ready then delete those 10 containers) against this PR and against master. I got this results:
master:
```
real 125.61
user 1.54
sys 0.33
```
This PR:
```
real 8.20
user 1.96
sys 0.38
```

I ran the runtime tests multiple times to ensure I didn't see any race conditions nor deadlocks (with and without `--race` flag).

tgraf commented on tests/07-performance.sh:
```diff
@@ -0,0 +1,76 @@
+function cleanup {
+    cd "${dir}/.."
+	git checkout "${current_branch}"
+	git stash pop
```
indent

tgraf commented on tests/07-performance.sh:
```diff
@@ -0,0 +1,76 @@
+function install_master {
+    cd "${dir}/.."
+    git stash
+    git checkout origin/master
```
This is pretty brutal. I totally get where you want to go an this data is very useful. I'm not sure we can stash/pop an just bluntly run make install in a test. People may run this outside of vagrant inside their dev environment. I'll think about this.

aanm replied:
Yeah, I got problems myself when there was nothing to stash and when cleaning up I was blindly doing `git stash pop`

aanm committed:
pkg/policy: changed Consumable LabelList to []*Labels

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
labels: fixed broken labels_test.go

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
mac: fixed broken mac_test.go

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
option: fixed broken option_test.go

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
pkg/option: add internal mutex to option.go

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
bpf/bpffs.go Use sync.Once for map lockdown

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
labels: add mutex to absKey

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
daemon: implemented GetCachedMaxLabelID

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
Introducing endpoints parallelization build

With this commit it is introduced a parallelization for the endpoints
regeneration. The daemon will have worker threads as the same number of
CPUs available on the system. Each of those worker threads will
signalize the endpoints turn to build itself.

Thus, the endpoints creation/regeneration will be asynchronous where the
user should verify if the endpoint had a successful regeneration with
`cilium endpoint list`.

Due the parallelization created around the endpoints builder there was a
need to create a mutex on various structures. Those mutex should be used
whenever that same structure's values are read or written.
All private methods of those structures will assume the structure locks
are locked, all public methods will assmue those structures locks are
unlocked, except when stated otherwise such as "MethodLocked", where
the public method needs to be called with the lock locked.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
daemon: made ConsumableCache concurrent aware

Add a RWMutex to the ConsumableCache structure so it can be used by
multiple go routines.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
bpf: add sync.Once to map Open

Signed-off-by: André Martins <andre@cilium.io>

tgraf commented on pkg/option/option.go:
```diff
@@ -121,6 +122,7 @@ func (om OptionMap) DeepCopy() OptionMap {
 }
 
 type BoolOptions struct {
+	optsMU  sync.RWMutex
```
👍

tgraf commented on pkg/bpf/bpffs.go:
```diff
@@ -29,8 +30,13 @@ var (
 
 	// Set to true on first get request to detect misorder
 	lockedDown = false
+	once       sync.Once
```
Did you run into a race with the existing code? I wonder whether it makes sense to start using abool instead.
https://github.com/tevino/abool

I assume that setting the boolean will not include any memory barriers so if other go subroutines rely on this it would still race. The existing code was not assuming parallelism but just coding mistakes.

aanm replied:
Are you talking of the race conditions between line 41 and the others like 48 for example?

tgraf replied:
The whole point of `lockedDown` is to catch whether the map root prefix is changed after the map root has been read. My question is whether you saw an actual race because just using `once` doesn't change the fact that this will still not have memory barriers around the access of `lockedDown` os if there was a real race, this would not fix it.

aanm replied:
Addressed

tgraf commented on daemon/docker_watcher.go:
```diff
@@ -280,9 +270,11 @@ func (d *Daemon) handleCreateContainer(id string, retry bool) {
-		containerCopy, ok := d.retrieveWorkingContainerCopy(id)
+		d.containersMU.RLock()
+		containerCopy, ok := d.containers[id]
+		d.containersMU.RUnlock()
```
We should rename the variable as it is not a copy anymore. Can be done in a follow-up commit though.

aanm replied:
It was fixed on a later commit, should I fix this on this commit?

tgraf replied:
That's fine then.

tgraf commented on daemon/docker_watcher.go:
```diff
@@ -280,9 +270,11 @@ func (d *Daemon) handleCreateContainer(id string, retry bool) {
+		d.containersMU.RUnlock()
 		if ok {
-			if orchLabelsModified = updateOrchLabels(&containerCopy, lbls); !orchLabelsModified {
+			if orchLabelsModified = updateOrchLabels(containerCopy, lbls); !orchLabelsModified {
```
How is it guaranteed that only one go subroutine modifies the container object via `updateOrchLabels()`?

aanm replied:
I made a whole bunch of changes on this PR and afterwards I split them into logic commits. The `containerCopy` is not guarantee that only a go subroutine modifies the container object at this time. I can move the `d.containerMU.RUnlock()` after the `updateOrchLabels` as a `Unlock()` on this commit. That way there won't be "bad" commits, sounds good?

tgraf replied:
Sounds good.

tgraf commented on pkg/policy/rule_consumers.go:
```diff
@@ -112,7 +112,7 @@ func (a *AllowRule) Allows(ctx *SearchContext) ConsumableDecision {
 type RuleConsumers struct {
 	Coverage []*labels.Label `json:"coverage,omitempty"`
-	Allow    []AllowRule     `json:"allow"`
+	Allow    []*AllowRule    `json:"allow"`
```
What is the motivation behind this change?

aanm replied:
`AllowRule` contains a `Label` which contains a mutex. `go vet` complains that the mutex pointer is being copied if the `AllowRule` is not a pointer.

tgraf commented on pkg/labels/labels.go:
```diff
@@ -98,8 +99,9 @@ type Label struct {
-	Source string `json:"source"`
-	absKey string
+	Source   string `json:"source"`
+	absKeyMU sync.RWMutex
```
Can you explain the case where this mutex is needed? Is the Label shared between multiple rules?

aanm replied:
There was a race condition, the `AbsoluteKey()` is called in multiple place like policy check, that can happen in parallel.

tgraf commented on daemon/ct.go:
```diff
@@ -65,21 +65,23 @@ func (d *Daemon) EnableConntrackGC() {
 					continue
 				}
-
+				e.Mutex.RUnlock()
```
Why is it safe to continue accessing the endpoint structure after unlocking the mutex? Are you just using the mutex to protect the Consumable pointer? If so, please document it.

aanm replied:
I add this:
```
// We can unlock the endpoint mutex sense
// in runGC it will be locked as needed.
```
Sounds sufficient?

tgraf replied:
The comment doesn't make sense to me because I don't see `runGC` taking any locks. I assuming accessing `e.ID` and calling `e.LogStatus` is fine without a lock is fine as we are holding `d.endpointsMU` so endpoints can't disappear. I think overall what will help to put a comment to every mutex to describe what the mutex is supposed to protect exactly. Is `e.Mutex` is supposed to protect writes to endpoints but reads are fine?

aanm replied:
Yes, that's true. So usually, mutexes are put before the variable they are protecting. That's something missing in the our mutexes. The endpoint ID is being written once, and that happens inside the endpoint with `NewEndpointFromChangeModel`, we don't need a mutex for the ID because of that. The `e.LogStatus` can be called because it's a public method and all public methods can be called since the lock will be locked by the receiver's function.

I'll move all mutex to the proper places and add a comment regarding which variables are they protecting.

tgraf commented on daemon/config.go:
```diff
@@ -38,6 +38,7 @@ type Config struct {
 	BpfDir                string                  // BPF template files directory
 	LibDir                string                  // Cilium library files directory
 	RunDir                string                  // Cilium runtime directory
+	LXCMapMU              sync.Mutex              // LXCMapMU is the mutex to access the LXCMap FIXME move this to the LXCMap struct?
```
Yes, please move to LXCMap struct. I think we should even move it into the generic BPF map abstraction.

aanm replied:
Moved to LXCMap struct. Created an issue to have a mutex in the generic BPF map abstraction.

tgraf commented on daemon/daemon.go:
```diff
@@ -80,6 +81,9 @@ type Daemon struct {
 	maxCachedLabelIDMU sync.RWMutex
 	maxCachedLabelID   policy.NumericIdentity
+	buildEndpointChan  chan *endpoint.Request
+	uniqueID           map[uint16]bool
```
We should stop hardcoding uint16 everyone. It will be very painful to undo. Let's just use uint64 here to never run into any issues. The build ID must not directly translate to a uint16 datapath ID.

aanm replied:
Addressed

tgraf commented on daemon/daemon.go:
```diff
@@ -88,13 +92,69 @@ func (d *Daemon) GetProxy() *proxy.Proxy {
+
+// EndpointBuilder creates #CPUs go routines that listen on the
+// `d.buildEndpointChan` for new endpoints.
+func (d *Daemon) EndpointBuilder() {
```
Can we call this `StartEndpointBuilders()`?

aanm replied:
Addressed

tgraf commented on daemon/daemon.go:
```diff
@@ -88,13 +92,69 @@ func (d *Daemon) GetProxy() *proxy.Proxy {
+func (d *Daemon) EndpointBuilder() {
+	log.Debugf("Creating %d worker threads", runtime.NumCPU())
+	// Create the same amount of worker threads as there are CPUs
+	for w := 0; w < runtime.NumCPU(); w++ {
```
This should be configurable as well.

tgraf replied:
Although I'm fine leaving it as-is for an initial version.

aanm replied:
Addressed

tgraf commented on daemon/daemon.go:
```diff
@@ -88,13 +92,69 @@ func (d *Daemon) GetProxy() *proxy.Proxy {
+				d.uniqueIDMU.Lock()
+				if isBuilding := d.uniqueID[e.ID]; isBuilding {
+					// Delete the endpoint from the queue
+					// only if was marked as isBuilding
```
The comment states what the code does which is obvious, why is it only deleting the entry if it was marked? In what case is it not marked given we do `d.uniqueID[e.ID] = true` just above?

aanm replied:
```
// In a case where the same endpoint enters the
// building queue, while it was still being build,
// it will be marked as `false`/"not building",
// thus, we only delete the endpoint from the
// queue only if it is marked as isBuilding.
```
Better?

tgraf commented on daemon/daemon.go:
```diff
@@ -88,13 +92,69 @@ func (d *Daemon) GetProxy() *proxy.Proxy {
+// QueueEndpoint puts the given request in the endpoints queue for processing.
+// The given request will receive 'true' in the MyTurn channel whenever it's its
+// turn or false if the request was denied/canceled.
+func (d *Daemon) QueueEndpoint(req *endpoint.Request) {
```
Can we rename this to `QueueEndpointBuild()`?

aanm replied:
Addressed

tgraf commented on pkg/labels/labels.go.orig:
```diff
@@ -0,0 +1,499 @@
+// Copyright 2016-2017 Authors of Cilium
```
Unwanted addition to the commit

aanm replied:
Addressed

tgraf commented on pkg/endpoint/policy.go:
```diff
@@ -311,41 +309,61 @@ func (e *Endpoint) regenerate(owner Owner) error {
+		ExternalDone: make(chan bool),
+	}
+	owner.QueueEndpoint(newReq)
+	go func(e *Endpoint, myTurn <-chan bool, finish chan<- bool, externalFinish chan<- bool) {
```
Should this goroutine be cancelled if the endpoint is deleted before the build happens?

aanm replied:
Nice catch, I changed the `StartEndpointBuilders` code a little bit:
```
...
go func() {
			for e := range d.buildEndpointChan {
				d.uniqueIDMU.Lock()
+				if _, ok := d.uniqueID[e.ID]; !ok {
+					// If the request is not present in the uniqueID,
+					// it means the request was deleted from the queue
+					// so we deny the request's turn.
+					e.MyTurn <- false
+					d.uniqueIDMU.Unlock()
+					continue
+				}
				// Set the endpoint to "building" state
				d.uniqueID[e.ID] = true
				e.MyTurn <- true
				d.uniqueIDMU.Unlock()
```

This way the goroutine sees that its turn was denied and ends cleanly.

aanm committed:
lxcmap: add mutex around the map

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
Moved all mutexes above the variables they protect

Since this is a go pattern we decided to move all mutexes for easy
reading.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
NEWS.rst: added the major changes for this PR

Signed-off-by: André Martins <andre@cilium.io>
